File: 979. Distribute Coins in Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
// How many coins need to pass through one node depends on given node's child node. So we can consider DFS to get the leaf node requirements, then populate to upper layer.
var distributeCoins = function(root) {
    let result = 0;
    (function dfs(node) {
        if (node) {
            const left = dfs(node.left);
            const right = dfs(node.right);
            result += Math.abs(left) + Math.abs(right);
            return node.val + left + right - 1; // Each node will need or give out Math.abs(node.val - 1) coins. 
        }
        return 0;
    })(root);
    return result;
};

File: Golang/Tree/979.distribute-coins-in-binary-tree.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Post order traversal
func distributeCoins(root *TreeNode) int {
	ret := 0
	abs := func(value int) int {
		return int(math.Abs(float64(value)))
	}
	var dfs func(node *TreeNode) int
	dfs = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		left := dfs(node.Left)
		right := dfs(node.Right)
		ret += abs(left) + abs(right)
		return node.Val + left + right - 1
	}
	return ret
}
